display_list/display_list_path_effect_unittests.cc

```diff
@@ -0,0 +1,223 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/display_list/display_list_attributes_testing.h"
+#include "flutter/display_list/display_list_builder.h"
+#include "flutter/display_list/display_list_comparable.h"
+#include "flutter/display_list/types.h"
+#include "gtest/gtest.h"
+#include "include/core/SkPath.h"
+#include "include/core/SkScalar.h"
+
+namespace flutter {
+namespace testing {
+
+TEST(DisplayListPathEffect, BuilderSetGet) {
+  const SkScalar TestDashes1[] = {4.0, 2.0};
+  auto dash_path_effect = DlDashPathEffect::Make(TestDashes1, 2, 0.0);
+  DisplayListBuilder builder;
+  ASSERT_EQ(builder.getPathEffect(), nullptr);
+  builder.setPathEffect(dash_path_effect.get());
+  ASSERT_NE(builder.getPathEffect(), nullptr);
+  ASSERT_TRUE(Equals(builder.getPathEffect(),
+                     static_cast<DlPathEffect*>(dash_path_effect.get())));
+  builder.setPathEffect(nullptr);
+  ASSERT_EQ(builder.getPathEffect(), nullptr);
+}
+
+TEST(DisplayListPathEffect, FromSkiaNullPathEffect) {
+  std::shared_ptr<DlPathEffect> path_effect = DlPathEffect::From(nullptr);
+  ASSERT_EQ(path_effect, nullptr);
+  ASSERT_EQ(path_effect.get(), nullptr);
+}
+
+TEST(DisplayListPathEffect, FromSkiaPathEffect) {
+  const SkScalar TestDashes2[] = {1.0, 1.5};
+  sk_sp<SkPathEffect> sk_path_effect =
+      SkDashPathEffect::Make(TestDashes2, 2, 0.0);
+  std::shared_ptr<DlPathEffect> dl_path_effect =
+      DlPathEffect::From(sk_path_effect);
+
+  ASSERT_EQ(dl_path_effect->type(), DlPathEffectType::kDash);
+  ASSERT_TRUE(
+      Equals(dl_path_effect, DlDashPathEffect::Make(TestDashes2, 2, 0.0)));
+}
+
+TEST(DisplayListPathEffect, EffectShared) {
+  const SkScalar TestDashes2[] = {1.0, 1.5};
+  auto effect = DlDashPathEffect::Make(TestDashes2, 2, 0.0);
+  ASSERT_TRUE(Equals(effect->shared(), effect));
+}
+
+TEST(DisplayListPathEffect, DashEffectAsDash) {
+  const SkScalar TestDashes2[] = {1.0, 1.5};
+  auto effect = DlDashPathEffect::Make(TestDashes2, 2, 0.0);
+  ASSERT_NE(effect->asDash(), nullptr);
+  ASSERT_EQ(effect->asDash(), effect.get());
+}
+
+TEST(DisplayListPathEffect, DashEffectEquals) {
+  const SkScalar TestDashes2[] = {1.0, 1.5};
+  auto effect1 = DlDashPathEffect::Make(TestDashes2, 2, 0.0);
+  auto effect2 = DlDashPathEffect::Make(TestDashes2, 2, 0.0);
+  ASSERT_TRUE(Equals(effect1, effect2));
+  ASSERT_TRUE(Equals(effect1->shared(), effect2->shared()));
+}
+
+TEST(DisplayListPathEffect, BlurNotEquals) {
+  const SkScalar TestDashes1[] = {4.0, 2.0};
+  const SkScalar TestDashes2[] = {1.0, 1.5};
```

Comment:
If you had a test where the first dash was equal and the second dash was not, you'd have discovered that your memcmp was using the wrong byte count.

Please add new dash arrays to test "interval 1 is different" and "interval 2 is different" and another that has 3 values (the first 2 of which match TestDashes1, but the 3rd is additional) so that you can test a difference in the interval count.